ridge-changemap.1
.TH ridge-changemap 1 "May 7, 2013" "Surrey Space Centre" 1.0
.SH NAME
ridge-changemap - Ridge-based urban change map generator
.SH SYNOPSIS
.B ridge-changemap
[\fIOPTION\fR ...] [\fB-m\fR \fIMODE\fR] \fICRDG\fR \fIPRE\fR
\fIPOST\fR \fIOUTFILE\fR
.SH DESCRIPTION
.PP
\fBridge-changemap\fR is a tool for generating change maps from SAR
amplitude images, with a building-by-building level of detail.  It is
designed for working with metre-resolution SAR images, such as those
available from the TerraSAR-X and COSMO-SkyMed platforms in high
resolution spotlight modes.  The algorithm and underlying change
metric used by this tool are described in [BRETT2012].
.PP
The tool takes three inputs.  \fICRDG\fR must contain ridge line data,
extracted using \fBridgetool\fR(1)'s `\fB-l\fR' mode, and usually
classified using \fBridge-classarea\fR(1) to indentify features likely
to be associated with buildings.  \fIPRE\fR must be the pre-event SAR
image used to generate \fICRDG\fR, and \fIPOST\fR must be the
corresponding post-event image, coregistered with \fIPRE\fR.  Both
TIFF files must be in single-channel, 32-bit floating point format.
.PP
Output is generated in \fIOUTFILE\fR, depending on the selected
\fIMODE\fR.  The output format is detected from the name of
\fIOUTFILE\fR, and may either be Portable Document Format (PDF) or
Portable Network Graphics (PNG).
.SH MODES
.PP The tool supports two rendering modes for the generated change map:
.TP 8
\fBridgelines\fR
The curvilinear features from the ridge file are drawn as vectors,
with each line segment coloured according to the level of change
detected.  This is the default mode.
.TP 8
\fBridgemask\fR
A raster image is created, and each pixel is coloured according to
detected level of change only if intersected by a curvilinear feature.
This is the approach described in [BRETT2012].
.SH OPTIONS
.TP 8
\fB-m\fR, \fB--mode\fR=\fIMODE\fR
Select the rendering mode used for generating the output image.  The
possible \fIMODE\fRs are described above.
.TP 8
\fB-c\fR, \fB--class\fR=\fICLASS\fR
For classified ridge data files, set the desired class label to
\fICLASS\fR.  All curvilinear features without this class label is
discarded.  If the ridge data file does not contain classified data,
all data is used and this option has no effect.
.TP 8
\fB-i\fR, \fB--nan\fR=\fIVAL\fR
Replace bad pixel values found in the input SAR images with
\fIVAL\fR. Because some remote sensing image products contain
non-finite or NaN pixel values, \fBridge-changemap\fR will replace
these with a finite value before change map generation.  By default,
the replacement value is 0.
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.SH REFERENCES
.TP 8
[BRETT2012]
Brett, Peter T.B.; Guida, Raffaella, \fIGeometry-based SAR curvilinear
feature selection for damage detection\fR, Synthetic Aperture Radar,
2012. EUSAR. 9th European Conference on, pp.62,65, 23-26
April 2012
.SH AUTHOR
Written by Peter TB Brett, Surrey Space Centre, University of Surrey,
UK.
.SH COPYRIGHT
.PP
Copyright \(co 2013 Peter TB Brett <p.brett@surrey.ac.uk>.  License
GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
\fBridgetool\fR(1), \fBridge-classarea\fR(1).


ridge-changemap.c
/*
 * Surrey Space Centre urban change detection tool for SAR
 * Copyright (C) 2013 Peter Brett <p.brett@surrey.ac.uk>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <errno.h>
#include <getopt.h>
#include <assert.h>

#include <glib.h>
#include <ridgeutil.h>
#include <ridgeio.h>

#include "ridge-changemap.h"

#define DEFAULT_CLASS_LABEL 1

enum OutputMode {
  MODE_RIDGE_LINES,
  MODE_RIDGE_MASK,
};

/* -------------------------------------------------------------------- */

#define GETOPT_OPTIONS "c:hi:m:"

struct option long_options[] =
  {
    {"class", 1, 0, 'c'},
    {"help", 0, 0, 'h'},
    {"mode", 1, 0, 'm'},
    {"nan", 1, 0, 'i'},
    {0, 0, 0, 0} /* Guard */
  };

static void
usage (char *name, int status)
{
  printf (
"Usage: %s [OPTION ...] [-m MODE] CRDG PRE POST OUTFILE\n"
"\n"
"Modes:\n"
"  ridgelines      Draw vector features coloured by change\n"
"  ridgemask       Draw masked ratio image coloured by change\n"
"\n"
"Options:\n"
"  -m, --mode=MODE Set changemap rendering mode [ridgelines]\n"
"  -c, --class=CLASS  Set class label to use for detection [%i]\n"
"  -i, --nan=VAL   Set non-finite input values to VAL [default 0]\n"
"  -h, --help      Display this message and exit\n"
"\n"
"Generates a change map using a pre-event SAR amplitude image PRE, a\n"
"post-event image POST, and a classified ridge data file CRDG.  Output\n"
"is generated in OUTFILE.  All images should be single-channel 32-bit\n"
"floating point TIFF files.\n"
"\n"
"Please report bugs to %s.\n",
name, DEFAULT_CLASS_LABEL, PACKAGE_BUGREPORT);
  exit (status);
}

/* -------------------------------------------------------------------------- */

static RioData *
ridges_load_check (const char *crdg_fn, int class_label,
                   uint32_t *height, uint32_t *width)
{
  g_assert (crdg_fn);
  g_assert (height);
  g_assert (width);

  RioData *data = rio_data_from_file (crdg_fn);
  if (data == NULL) {
    fprintf (stderr, "ERROR: Failed to load ridge data from '%s': %s.\n",
             crdg_fn, strerror (errno));
    exit (2);
  }

  if (rio_data_get_type (data) != RIO_DATA_LINES) {
    fprintf (stderr, "ERROR: '%s' does not contain ridge line data.\n",
             crdg_fn);
    exit (2);
  }

  /* Determine original image height & width */
  if (!(rio_data_get_metadata_uint32 (data, RIO_KEY_IMAGE_ROWS, height)
        && rio_data_get_metadata_uint32 (data, RIO_KEY_IMAGE_COLS, width))) {
    fprintf (stderr, "ERROR: '%s' contains invalid image size metadata.\n",
             crdg_fn);
    exit (2);
  }

  size_t N = rio_data_get_num_entries (data);

  /* Load classification metadata */
  size_t Nc;
  uint8_t *classification =
    (uint8_t *) rio_data_get_metadata (data, RIO_KEY_IMAGE_CLASSIFICATION, &Nc);

  if (classification == NULL || Nc != N) {
    fprintf (stderr, "WARNING: '%s' contains invalid classification metadata.\n",
             crdg_fn);

    /* If no classification data is present, just return the original
     * ridge data set. */
    return data;
  }

  /* Create a new ridge data set, and copy over all of the features
   * that have the correct class label. */
  RioData *data_c = rio_data_new (RIO_DATA_LINES);
  rio_data_set_metadata_uint32 (data_c, RIO_KEY_IMAGE_ROWS, *height);
  rio_data_set_metadata_uint32 (data_c, RIO_KEY_IMAGE_COLS, *width);
  for (int i = 0; i < N; i++) {
    if (classification[i] != class_label) continue;

    RioLine *l = rio_data_get_line (data, i);
    RioLine *l_c = rio_data_new_line (data_c);
    int M = rio_line_get_length (l);
    for (int j = 0; j < M; j++) {
      RioPoint *p = rio_line_get_point (l, j);
      RioPoint *p_c = rio_line_new_point (l_c);
      memcpy (p_c, p, sizeof(RioPoint));
    }
  }

  rio_data_destroy (data);
  return data_c;
}

static RutSurface *
img_load_check (const char *fn, uint32_t rows, uint32_t cols)
{
  RutSurface *img = rut_surface_from_tiff (fn);
  if (img == NULL) {
    fprintf (stderr, "ERROR: Failed to load TIFF from '%s'.\n", fn);
    exit (3);
  }
  /* Check size */
  if (img->rows != rows || img->cols != cols) {
    fprintf (stderr, "ERROR: Bad image size for '%s' (expected %ux%u).\n", fn,
             rows, cols);
    exit (3);
  }
  return img;
}

/* -------------------------------------------------------------------------- */

int
guess_output_format (const char *filename)
{
  struct _FormatSuffix {
    const char *suffix;
    int format;
  };
  struct _FormatSuffix format_suffixes[] = {
    {"png", FORMAT_PNG},
    {"pdf", FORMAT_PDF},
    {NULL, FORMAT_NONE},
  };

  const char *out_suffix = strrchr (filename, '.');
  if (out_suffix != NULL) {
    ++out_suffix; /* Skip '.' */
    for (struct _FormatSuffix *s = format_suffixes; s->suffix != NULL; ++s) {
      if (strcasecmp (out_suffix, s->suffix) == 0) {
        return s->format;
      }
    }
  }
  return FORMAT_NONE;
}

/* -------------------------------------------------------------------------- */

int
main (int argc, char **argv)
{
  int c, status;
  int cfg_mode = MODE_RIDGE_LINES;
  uint8_t cfg_class = DEFAULT_CLASS_LABEL;
  double cfg_nan = 0;
  int cfg_smooth = 0;
  char *cfg_crdg_fn = NULL;
  char *cfg_pre_fn = NULL;
  char *cfg_post_fn = NULL;
  char *cfg_out_fn = NULL;
  int cfg_format = FORMAT_NONE;

  while (1) {
    c = getopt_long (argc, argv, GETOPT_OPTIONS, long_options, NULL);
    if (c == -1) break;

    switch (c) {
    case 'c':
      status = sscanf (optarg, "%hhu", &cfg_class);
      if (status != 1) {
        fprintf (stderr, "ERROR: Bad argument '%s' to -c option.\n\n",
                 optarg);
        usage (argv[0], 1);
      }
      break;
    case 'h':
      usage (argv[0], 0);
      break;
    case 'i':
      status = sscanf (optarg, "%lf", &cfg_nan);
      if (status != 1) {
        fprintf (stderr, "ERROR: Bad argument '%s' to -i option.\n\n",
                 optarg);
        usage (argv[0], 1);
      }
      break;
    case 'm':
      if (strcmp (optarg, "ridgelines") == 0) {
        cfg_mode = MODE_RIDGE_LINES;
      } else if (strcmp (optarg, "ridgemask") == 0) {
        cfg_mode = MODE_RIDGE_MASK;
      } else {
        fprintf (stderr, "ERROR: Bad argument '%s' to -m option.\n\n",
                 optarg);
        usage (argv[0], 1);
      }
      break;
    case 's':
      cfg_smooth = 1;
      break;

    case '?':
      usage (argv[0], 1);
    default:
      g_assert_not_reached ();
    }
  }

  /* Get filenames */
  if (argc - optind < 4) {
    fprintf (stderr,
             "ERROR: You must specify a ridge data file, pre- and post-event SAR\n"
             "images, and an output filename.\n\n");
    usage (argv[0], 1);
  }

  cfg_crdg_fn = argv[optind++];
  cfg_pre_fn = argv[optind++];
  cfg_post_fn = argv[optind++];
  cfg_out_fn = argv[optind++];

  /* Initialise change map structure */
  ChangeMap *changes = change_map_new ();
  change_map_set_nan (changes, cfg_nan);

  /* Load & check ridge data */
  uint32_t height, width;
  RioData *ridges = ridges_load_check (cfg_crdg_fn, cfg_class, &height, &width);
  change_map_set_ridge_data (changes, ridges);

  /* Load & check pre/post SAR images */
  RutSurface *pre, *post;
  pre = img_load_check (cfg_pre_fn, height, width);
  change_map_set_pre_image (changes, pre);
  post = img_load_check (cfg_post_fn, height, width);
  change_map_set_post_image (changes, post);

  /* Figure out desired output file format */
  /* FIXME should be an explicit command-line option */
  if (cfg_format == FORMAT_NONE) {
    cfg_format = guess_output_format (cfg_out_fn);
  }
  if (cfg_format == FORMAT_NONE) {
    fprintf (stderr, "WARNING: Could not guess output format for '%s'. Using PDF.\n",
             cfg_out_fn);
    cfg_format = FORMAT_PDF;
  }

  /* Output! */
  OutputOptions export_opts;
  export_opts.filename = cfg_out_fn;
  export_opts.format = cfg_format;
  export_opts.height = height;
  export_opts.width = width;

  switch (cfg_mode) {
  case MODE_RIDGE_LINES:
    export_ridge_lines (changes, &export_opts);
    break;
  case MODE_RIDGE_MASK:
    export_ridge_mask (changes, &export_opts);
    break;
  default:
    g_assert_not_reached ();
  };

  /* Cleanup */
  change_map_free (changes);
  rio_data_destroy (ridges);
  rut_surface_destroy (pre);
  rut_surface_destroy (post);
  return 0;
}
